src/System.Net.Http/src/uap/System/Net/HttpHandlerToFilter.cs

```diff
@@ -26,65 +26,207 @@ namespace System.Net.Http
+        private readonly RTHttpBaseProtocolFilter _filter;
+        private readonly RTHttpBaseProtocolFilter _filterWithNoCredentials;
         private int _filterMaxVersionSet;
+        private HttpClientHandler _handler;
 
-        internal HttpHandlerToFilter(RTHttpBaseProtocolFilter filter)
+        internal HttpHandlerToFilter(
+            RTHttpBaseProtocolFilter filter,
+            RTHttpBaseProtocolFilter filterWithNoCredentials,
+            HttpClientHandler handler)
         {
             if (filter == null)
             {
                 throw new ArgumentNullException(nameof(filter));
             }
 
-            _next = filter;
+            _filter = filter;
+            _filterWithNoCredentials = filterWithNoCredentials;
             _filterMaxVersionSet = 0;
+            _handler = handler;
         }
 
         internal string RequestMessageLookupKey { get; set; }
         internal string SavedExceptionDispatchInfoLookupKey { get; set; }
 
         protected internal override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancel)
         {
+            int redirects = 0;
+            HttpMethod requestHttpMethod;
+            bool skipRequestContentIfPresent = false;
+            HttpResponseMessage response = null;
+
             if (request == null)
             {
                 throw new ArgumentNullException(nameof(request));
             }
-            cancel.ThrowIfCancellationRequested();
 
-            RTHttpRequestMessage rtRequest = await ConvertRequestAsync(request).ConfigureAwait(false);
+            requestHttpMethod = request.Method;
 
-            RTHttpResponseMessage rtResponse;
-            try
-            {
-                rtResponse = await _next.SendRequestAsync(rtRequest).AsTask(cancel).ConfigureAwait(false);
-            }
-            catch (TaskCanceledException)
-            {
-                throw;
-            }
-            catch (Exception)
+            while (true)
             {
-                object info;
-                if (rtRequest.Properties.TryGetValue(SavedExceptionDispatchInfoLookupKey, out info))
+                cancel.ThrowIfCancellationRequested();
+
+                if (response != null)
                 {
-                    ((ExceptionDispatchInfo)info).Throw();
+                    response.Dispose();
+                    response = null;
                 }
 
-                throw;
-            }
+                RTHttpRequestMessage rtRequest = await ConvertRequestAsync(
+                    request,
+                    requestHttpMethod,
+                    skipRequestContentIfPresent).ConfigureAwait(false);
+
+                RTHttpResponseMessage rtResponse;
+                try
+                {
+                    if (redirects > 0)
+                    {
+                        rtResponse = await _filterWithNoCredentials.SendRequestAsync(rtRequest).AsTask(cancel).ConfigureAwait(false);
```

Comment:
Could we create the filter right here as a clone of the original filter but leaving out the credentials (instead of keeping 2 filter objects just in case there is a redirection - which in most cases I believe is not true).